prow/git/v2/client_factory.go

```diff
@@ -41,37 +43,107 @@ type ClientFactory interface {
 type RepoClient interface {
 	Publisher
 	Interactor
+	Executor
 }
 
 type repoClient struct {
 	publisher
 	interactor
 }
 
-// NewClientFactory allows for the creation of repository clients
-func NewClientFactory(host string, useSSH bool, username LoginGetter, token TokenGetter, gitUser GitUserGetter, censor Censor) (ClientFactory, error) {
-	cacheDir, err := ioutil.TempDir("", "gitcache")
+type ClientFactoryOpts struct {
+	// Host, defaults to "github.com" if unset
+	Host string
+	// UseSSH, defaults to false
+	UseSSH *bool
+	// The directory in which the cache should be
+	// created. Defaults to the "/var/tmp" on
+	// Linux and os.TempDir otherwise
+	CacheDirBase *string
+	// If unset, publishing action will error
+	Username LoginGetter
+	// If unset, publishing action will error
+	Token TokenGetter
+	// The git user to use.
+	GitUser GitUserGetter
+	// The censor to use. Not needed for anonymous
+	// actions.
+	Censor Censor
+}
+
+// Apply allows to use a ClientFactoryOpts as Opt
+func (cfo *ClientFactoryOpts) Apply(target *ClientFactoryOpts) {
+	if cfo.Host != "" {
+		target.Host = cfo.Host
+	}
+	if cfo.UseSSH != nil {
+		target.UseSSH = cfo.UseSSH
+	}
+	if cfo.CacheDirBase != nil {
+		target.CacheDirBase = cfo.CacheDirBase
+	}
+	if cfo.Token != nil {
+		target.Token = cfo.Token
+	}
+	if cfo.GitUser != nil {
+		target.Token = cfo.Token
+	}
+	if cfo.Censor != nil {
+		target.Censor = cfo.Censor
+	}
+}
+
+// ClientFactoryOpts allows to manipulate the options for a ClientFactory
+type ClientFactoryOpt func(*ClientFactoryOpts)
+
+func defaultClientFactoryOpts(cfo *ClientFactoryOpts) {
+	if cfo.Host == "" {
+		cfo.Host = "github.com"
+	}
+	if cfo.CacheDirBase == nil {
+		switch runtime.GOOS {
+		case "linux":
+			cfo.CacheDirBase = utilpointer.StringPtr("/var/tmp")
```

Comment:
Doesn't Go have a tempdir function?

Reply:
Yeah we still use that, the problem is it returns `"/tmp"` on linux which is tmpfs backed. If this git client is used in a component that clones many, potentially big repos, this may result in gigabytes of data and either the laptop crashing or a `no space on device` error.

`"/var/tmp"` is intended for temporary files that may be bigger and might be retained longer: https://unix.stackexchange.com/questions/30489/what-is-the-difference-between-tmp-and-var-tmp

It even exists on mac